File: 6.cpp
/*
    struct ListNode{
        int val;
        struct ListNode *next;
        ListNode(int x):
            val(x), next(NULL){}
    }
*/

// 倒序打印链表
vector<int> printListFromTailToHead(ListNode* head){
    std::stack<int> list;
    std::vector<int> ans;
    while(head != NULL){
        list.push(head -> val);
        head = head -> next;
    }
    while(!list.empty()){
        ans.push_back(list.top());
        list.pop();
    }
    return ans;
}


// 递归形式
void printListFromTailToHead(ListNode* head){
    if(head -> next == NULL){
        printf("%d ", head -> val);
        return;
    }
    printListFromTailToHead(head -> next);
    printf("%d ", head -> val);
    return;
}

// 结尾增加节点
void addNode(ListNode*& head, int val){
    ListNode* node = new ListNode(val);
    ListNode* p;
    if(head == nullptr){
        head = node;
        return;
    }
    p = head;
    while(p -> next != nullptr){
        p = p -> next;
    }
    p -> next = node;
    return;
}

File: 6.java
import java.util.ArrayList;

public class Solution {

    public ArrayList<Integer> printListFromTailToHead(ListNode head) {

        ArrayList<Integer> arr = new ArrayList<Integer>();

        if (head == null) {
            return arr;
        }

        ListNode p = head;

        for (; p != null; p = p.next) {
            arr.add(p.val);
        }
       
        for (int i = 0; i < arr.size() / 2; ++i) {
            int temp = arr.get(i);
            arr.set(i, arr.get(arr.size() - i - 1));
            arr.set(arr.size() - i - 1, temp);
        }

        return arr;
    }
}

/*  使用栈

public class Solution {

    public ArrayList<Integer> printListFromTailToHead(ListNode head) {

        ArrayList<Integer> arr = new ArrayList<Integer>();
        Stack<Integer> stack = new Stack<Integer>();

        if (head == null) {
            return arr;
        }

        ListNode p = head;

        for (; p != null; p = p.next) {
            stack.push(p.val);
        }
       
        while (!stack.empty()) {
            arr.add(stack.pop());
        }

        return arr;
    }
}


*/

/*  使用递归

public class Solution {
    
    ArrayList<Integer> arr = new ArrayList<Integer>();

    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {

        if (listNode != null) {
            printListFromTailToHead(listNode.next);
            arr.add(listNode.val);
        }

        return arr;
    }
}


*/
